### PR TITLE
SONARJAVA-4406 FP on S2142 when the InterruptedException is caught in an inner try-catch

### PR DESCRIPTION
**Pull request by kaufco**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java [hunk 1]
```diff
@@ -148,9 +149,45 @@ private static boolean isFinallyInterrupting(@Nullable BlockTree blockTree) {
...
+        symbol -> symbol.thrownTypes().stream().anyMatch(INTERRUPTING_TYPE_PREDICATE)
+      );
+    }
+
+    @Override
+    public void visitTryStatement(TryStatementTree tryStatementTree) {
+      // If inner `try` statement catches interrupting types: cut analysis of its try block, because possible
+      // interruptions thrown there are handled within the scope of the catch block.
+      // Yet, all other blocks (resources, catch, finally) of inner `try`s must still be analyzed, because possible
+      // interruptions thrown there must be handled within the scope of the outer try.
```
Comment (leveretka) [inline java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java:169]: Nice comment

File: java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java [hunk 2]
```diff
@@ -148,9 +149,45 @@ private static boolean isFinallyInterrupting(@Nullable BlockTree blockTree) {
...
+      if (!isCatchingAnyInterruptingTypes) {
+        scan(tryStatementTree.block());
+      }
+      scan(tryStatementTree.catches());
+      scan(tryStatementTree.finallyBlock());
+    }
+
+    @Override
+    public void visitThrowStatement(ThrowStatementTree tree) {
+      // besides to method invocation, we also need to collect throw statements
```
Comment (leveretka) [inline java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java:185]: While you're right that we should fix it, I'd prefer if you split it into 2 separate commits and have a new ticket. You can do it in 1 PR, just 2 separate commits so we keep track of the issue correctly

File: java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java [hunk 3]
```diff
@@ -148,9 +149,45 @@ private static boolean isFinallyInterrupting(@Nullable BlockTree blockTree) {
...
+      // If inner `try` statement catches interrupting types: cut analysis of its try block, because possible
+      // interruptions thrown there are handled within the scope of the catch block.
+      // Yet, all other blocks (resources, catch, finally) of inner `try`s must still be analyzed, because possible
+      // interruptions thrown there must be handled within the scope of the outer try.
+
+      boolean isCatchingAnyInterruptingTypes = tryStatementTree.catches().stream().anyMatch(catchTree ->
+        getCaughtTypes(catchTree.parameter()).stream().anyMatch(INTERRUPTING_TYPE_PREDICATE)
+      );
+
+      scan(tryStatementTree.resourceList());
```
[the remark this answers is not in the document]
Comment (kaufco) [reply java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java:175]: Yes, it's required. Added another test case to show that (`throwIoExceptionFromResourceList`)

File: java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java [hunk 4]
```diff
@@ -148,9 +149,45 @@ private static boolean isFinallyInterrupting(@Nullable BlockTree blockTree) {
...
+      scan(tryStatementTree.finallyBlock());
+    }
+
+    @Override
+    public void visitThrowStatement(ThrowStatementTree tree) {
+      // besides to method invocation, we also need to collect throw statements
+      if (isInterruptingExceptionExpression(tree.expression())) {
+        invocationTree.add(tree);
+      }
+      super.visitThrowStatement(tree);
```
[the remark this answers is not in the document]
Comment (kaufco) [reply java-checks/src/main/java/org/sonar/java/checks/InterruptedExceptionCheck.java:189]: Yes, still needed, even though it's a very constructed corner case. Added test case `throwIoExceptionWithinThrowStatement` to demonstrate it.

**Review by leveretka — APPROVED** [review]
LGTM!
